crates/vim/src/normal.rs

```diff
@@ -141,7 +143,28 @@ pub(crate) fn register(workspace: &mut Workspace, cx: &mut ViewContext<Workspace
         Vim::update(cx, |vim, cx| {
             vim.record_current_action(cx);
             vim.update_active_editor(cx, |_, editor, cx| {
-                editor.transact(cx, |editor, cx| editor.indent(&Default::default(), cx))
+                editor.transact(cx, |editor, cx| {
+                    let mut original_positions: HashMap<_, _> = Default::default();
+                    let (map, selections) = editor.selections.all_display(cx);
+                    selections.iter().for_each(|selection| {
+                        let line_start = first_non_whitespace(&map, true, selection.start).column();
+                        original_positions.insert(selection.id, (selection.start, line_start));
+                    });
+                    editor.indent(&Default::default(), cx);
+                    editor.set_clip_at_line_ends(true, cx);
+                    editor.change_selections(Some(Autoscroll::fit()), cx, |s| {
+                        s.move_with(|map, selection| {
+                            let (mut cursor, original_line_start) =
+                                original_positions.remove(&selection.id).unwrap();
+                            *cursor.column_mut() +=
+                                first_non_whitespace(map, true, DisplayPoint::new(cursor.row(), 0))
+                                    .column()
+                                    - original_line_start;
+                            cursor = map.clip_point(cursor, Bias::Left);
+                            selection.collapse_to(cursor, selection.goal);
+                        });
+                    });
+                });
```

Comment:
```suggestion
                editor.transact(cx, |editor, cx| {
                    let (_, selections) = editor.selections.all_display(cx);
                    let mut original_positions = selections
                        .iter()
                        .map(|selection| (selection.id, selection.start))
                        .collect::<HashMap<_, _>>();
                    editor.indent(&Default::default(), cx);

                    editor.change_selections(Some(Autoscroll::fit()), cx, |s| {
                        s.move_with(|_, selection| {
                            if let Some(cursor) = original_positions.remove(&selection.id) {
                                selection.collapse_to(cursor, selection.goal);
                            }
                        });
                    });
                });
```

If we ignore the column, I think this is enough to make it work (it works for me).

Reply:
I predict a crash in rare edge cases involving multibyte characters :D.

Additionally, we should set SelectionGoal::None in this case. The selection goal represents the "ideal" column when moving up and down; when moving sideways it is always reset to None.